async/src/traits/consumer.rs

```diff
@@ -70,21 +70,28 @@ impl<'a, A: AsyncConsumer> Future for PopFuture<'a, A> {
     type Output = Option<A::Item>;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        assert!(!self.done);
-        self.owner.register_write_waker(cx.waker());
-        let closed = self.owner.is_closed();
-        #[cfg(feature = "std")]
-        std::println!("PopFuture::poll: closed={}", closed);
-        match self.owner.try_pop() {
-            Some(item) => {
-                self.done = true;
-                Poll::Ready(Some(item))
-            }
-            None => {
-                if closed {
-                    Poll::Ready(None)
-                } else {
-                    Poll::Pending
+        let mut waker_registered = false;
+        loop {
```

Comment:
Nice solution, checking ring buffer in a loop really eliminates the need to register waker in advance.